File: java_solutions/2605.form-smallest-number-from-two-digit-arrays.java
/*
 * @lc app=leetcode id=2605 lang=java
 *
 * [2605] Form Smallest Number From Two Digit Arrays
 */

class Solution {
    public int minNumber(int[] nums1, int[] nums2) {
        return IntStream.of(Arrays.stream(nums1).reduce(0, (m, v) -> m | (1 << v)))
            .map(m1 -> IntStream.of(Arrays.stream(nums2).reduce(0, (m, v) -> m | (1 << v)))
                .map(m2 -> (m1 & m2) > 0 ? Integer.numberOfTrailingZeros(m1 & m2) : 
                    Math.min(Integer.numberOfTrailingZeros(m1), Integer.numberOfTrailingZeros(m2)) * 10 + 
                    Math.max(Integer.numberOfTrailingZeros(m1), Integer.numberOfTrailingZeros(m2)))
                .findFirst().getAsInt())
            .findFirst().getAsInt();
    }
}

File: kotlin_solutions/2605.form-smallest-number-from-two-digit-arrays.kt
/*
 * @lc app=leetcode id=2605 lang=kotlin
 *
 * [2605] Form Smallest Number From Two Digit Arrays
 */

class Solution {fun minNumber(nums1: IntArray, nums2: IntArray) = nums1.toSet().intersect(nums2.toSet()).minOrNull()?.let { minOf(it, (minOf(nums1.min(), nums2.min()).toString() + maxOf(nums1.min(), nums2.min()).toString()).toInt()) } ?: (minOf(nums1.min(), nums2.min()).toString() + maxOf(nums1.min(), nums2.min()).toString()).toInt()}

